feat(app): use crc service

olevski opened the pull request:
Uses the compute resource access control service for launching sessions.

rokroskar commented on Dockerfile:
```diff
@@ -1,23 +1,29 @@
-
-FROM base as builder
-ENV POETRY_HOME=/opt/poetry
+FROM python:3.11-bullseye as builder
```
python 3.10 is used elsewhere - is that on purpose?

olevski replied:
@rokroskar the other components install the renku cli. And I think the CLI is still not fully compatible with python 3.11. This one does not use the CLI so we can go to 3.11.

rokroskar replied:
if you don't need 3.11 we could keep it consistent?

olevski replied:
Ok cool I will update to 3.11.

rokroskar replied:
oh it does? then we can build 3.11 base images as well?

olevski replied:
The problem I had was that I could not get numpy installed on the debian images I was using with python 3.11. But now since we removed numpy from the dependencies things work.

Panaetius commented on renku_notebooks/api/classes/crc.py:
```diff
@@ -0,0 +1,147 @@
+    ) -> Optional[ServerOptions]:
+        """Find a resource class that is available to the user that is greater than or equal to
+        the old-style server options that the user requested."""
+        headers = None
+        if user.access_token is not None:
+            headers = {"Authorization": f"bearer {user.access_token}"}
+        res = requests.get(self.crc_url + "/resource_pools", headers=headers)
+        if res.status_code != 200:
+            raise IntermittentError(
+                message="The compute resource access control service sent "
+                "an unexpected response, please try again later",
+            )
+        resource_pools = res.json()
```
This code is duplicated three times (here and above), maybe it should be its own method?

Panaetius commented on renku_notebooks/api/classes/server.py:
```diff
@@ -221,40 +222,21 @@ def _get_registry_secret(self, b64encode=True):
-                )
-            else:
-                resources["limits"]["cpu"] = cpu_request
+            resources["limits"]["cpu"] = 3 * cpu_request
```
What's the `3` here?

olevski replied:
This is to allow the cpu to spike up momentarily. Because if we set the limit on the cpu it gets throttled and then when you create a new session it takes a really long time for the session to start up. If the limits are set to "lax" (i.e. "relaxed") then the limit is larger than the request. But if the limits are set to strict then the limit is the same as the request and the cpu is throttled.

Panaetius replied:
makes sense. you should give it a descriptive name, random magic numbers in code aren't nice :slightly_smiling_face:

Panaetius commented on renku_notebooks/api/schemas/server_options.py:
```diff
@@ -1,82 +1,159 @@
+            cpu=data["cpu_request"],
+            gpu=data["gpu_request"],
+            memory=data["mem_request"],
+            default_url=data["defaultUrl"],
```
We have a weird mix of camel case and snake case in request fields, we should clean this up at some point

olevski replied:
I will open an issue for this. We should coordinate with the UI.

olevski replied:
https://github.com/SwissDataScienceCenter/renku-notebooks/issues/1525

Panaetius reviewed (changes requested):
thank you! nice to see this taking shape :slightly_smiling_face:

olevski commented on Dockerfile:
```diff
@@ -1,23 +1,29 @@
+WORKDIR /app
+RUN python3 -m pip install --user pipx && \
+    python3 -m pipx ensurepath && \
+    /home/renku/.local/bin/pipx install poetry && \
```
@Panaetius I use pipx because I want poetry to be in a separate environment and I do not want to have poetry at all in the final image.

So either I use pipx or I create a separate virtualenv for poetry and then install poetry there. Pipx was easier to use. So I will leave this as is.

olevski commented on git_services/Dockerfile.sidecar:
```diff
@@ -1,24 +1,30 @@
+WORKDIR /app
+RUN python3 -m pip install --user pipx && \
+    python3 -m pipx ensurepath && \
+    /home/jovyan/.local/bin/pipx install poetry && \
```
I will leave this as is as we discussed.

olevski commented on renku_notebooks/api/classes/crc.py:
```diff
@@ -0,0 +1,147 @@
+            raise ConfigurationError("Cannot find the default resource pool.")
+        default_pool = default_pools[0]
+        default_classes = [
+            cls for cls in default_pool.get("classes", []) if cls.get("default", False)
```
![image](https://github.com/SwissDataScienceCenter/renku-notebooks/assets/16360283/641ffb37-82b1-4e0b-a8e3-454e7728ab6a)

The classes field is a list. Not a dictionary.

Panaetius replied:
sorry my bad, I got mixed up with the `cls.get` after and thought it should be dict...

olevski commented on renku_notebooks/api/classes/crc.py:
```diff
@@ -0,0 +1,147 @@
+            raise InvalidComputeResourceError(
+                message="The requested storage surpasses the maximum value allowed."
+            )
+        # Memory and disk space in CRC are assumed to be in gigabytes whereas
+        # the notebook service assumes that if a plain number is used then it is bytes.
```
So the comment is not right. I updated the schema for the request in launch_notebook to be in gigabytes. But I did not update the comment. Now I added the assumption that disk and memory are in gigabytes in the docstring.

olevski commented:
@Panaetius I addressed all the comments. The integration tests are failing because it seems that the oauth application used by the tests has been removed from gitlab. I will fix this later. I ran the integration tests locally and they all passed.
